File: src/game/defense/defensePhraseJudge.ts
/**
 * Defense mode phrase judgment: sequential steps across N measures, completion count, pending switch.
 */
import type {
  DefenseAttackTrigger,
  DefensePhrase,
  DefensePhraseChord,
  DefensePlayStyle,
} from '@/game/defense/defenseTypes';
import { resolveDefenseStaffChordLabelForStep } from '@/game/defense/defenseStaffChordLabel';
import { parseProgressionChordRoot, rootMidiBelow } from '@/game/training/trainingProgression';
import {
  computeOrderedChordKeyboardHintsFromMidis,
  orderedPitchClassesFromMidis,
  shouldAcceptChordPitchClassInput,
  type OrderedChordKeyboardHints,
} from '@/utils/orderedChordInput';
import {
  buildExpectedPitchCandidates,
  type ExpectedPitchCandidates,
} from '@/utils/pitchInput/expectedPitchCandidates';
import {
  advanceChordStep,
  getPhraseChordSteps,
  type ChordStepAdvanceState,
  type PhraseChordStep,
} from '@/utils/phraseChordSteps';
import { normalizePitchClass } from '@/utils/phraseStreamMatching';

export interface DefensePhraseJudgeState {
  readonly phraseIndex: number;
  readonly chordIndex: number;
  readonly targetStepIndex: number;
  readonly correctNoteIndices: ReadonlySet<number>;
  readonly revealedNoteIndices: ReadonlySet<number>;
  readonly completionCount: number;
  readonly pendingSwitch: boolean;
  /** 直前に正解した pitch class。同音連打ゲート用。 */
  readonly lastAcceptedPitchClass: number | null;
}

interface DefensePhraseNoteEvaluation {
  readonly attack: boolean;
  readonly phraseCompleted: boolean;
  readonly measureCompleted: boolean;
  readonly pendingSwitch: boolean;
  readonly completionCount: number;
  readonly playRootMidi: number | null;
  readonly nextState: DefensePhraseJudgeState;
}

const emptySet = (): ReadonlySet<number> => new Set<number>();

export const createInitialPhraseJudgeState = (
  phraseIndex: number,
): DefensePhraseJudgeState => ({
  phraseIndex,
  chordIndex: 0,
  targetStepIndex: 0,
  correctNoteIndices: emptySet(),
  revealedNoteIndices: emptySet(),
  completionCount: 0,
  pendingSwitch: false,
  lastAcceptedPitchClass: null,
});

const getCurrentChord = (
  phrase: DefensePhrase | null,
  chordIndex: number,
): DefensePhraseChord | null => (
  phrase?.chords[chordIndex] ?? null
);

const resetChordProgress = (
  state: DefensePhraseJudgeState,
): DefensePhraseJudgeState => ({
  ...state,
  targetStepIndex: 0,
  correctNoteIndices: emptySet(),
  revealedNoteIndices: emptySet(),
});

const pitchClassFromMidi = (midi: number): number => normalizePitchClass(midi % 12);

const nextExpectedPitchClass = (
  phrases: readonly DefensePhrase[],
  state: DefensePhraseJudgeState,
  sequential: boolean,
): number | null => {
  const hints = getDefensePhraseKeyboardHints(phrases, state, sequential);
  if (hints.nextMidi == null) {
    return null;
  }
  return pitchClassFromMidi(hints.nextMidi);
};

export const isDefenseWaitingForSamePitchRepeat = (
  phrases: readonly DefensePhrase[],
  state: DefensePhraseJudgeState,
  sequential: boolean,
): boolean => {
  if (!sequential || state.lastAcceptedPitchClass === null) {
    return false;
  }
  const nextPc = nextExpectedPitchClass(phrases, state, sequential);
  return nextPc !== null && nextPc === state.lastAcceptedPitchClass;
};

const advanceChord = (
  state: DefensePhraseJudgeState,
  phrase: DefensePhrase,
): DefensePhraseJudgeState => {
  const chordCount = phrase.chords.length;
  if (chordCount === 0) {
    return resetChordProgress(state);
  }
  const nextIndex = (state.chordIndex + 1) % chordCount;
  return {
    ...state,
    chordIndex: nextIndex,
    targetStepIndex: 0,
    correctNoteIndices: emptySet(),
    revealedNoteIndices: emptySet(),
  };
};

const applyStepState = (
  state: DefensePhraseJudgeState,
  stepState: ChordStepAdvanceState,
): DefensePhraseJudgeState => ({
  ...state,
  targetStepIndex: stepState.targetStepIndex,
  correctNoteIndices: stepState.correctNoteIndices,
  revealedNoteIndices: stepState.revealedNoteIndices,
});

const requiredCompletionCountForPhrase = (
  phrase: DefensePhrase,
  stageDefault: number,
): number => phrase.requiredCompletionCount ?? stageDefault;

const stepMidiNotes = (
  chord: DefensePhraseChord,
  step: PhraseChordStep,
): number[] => {
  const midis: number[] = [];
  for (const noteIndex of step.noteIndices) {
    const note = chord.notes[noteIndex];
    if (note) {
      midis.push(note.pitchMidi);
    }
  }
  return midis;
};

const resolvePlayRootMidi = (
  chord: DefensePhraseChord,
  completedStepIndex: number,
  playRootOnChordChange: boolean,
): number | null => {
  if (!playRootOnChordChange) return null;
  const label = resolveDefenseStaffChordLabelForStep(chord, completedStepIndex);
  if (label.trim().length === 0) return null;
  const { steps } = getPhraseChordSteps(chord.notes);
  const step = steps[completedStepIndex];
  if (!step) return null;
  let lowestMidi = Number.POSITIVE_INFINITY;
  for (const noteIndex of step.noteIndices) {
    const note = chord.notes[noteIndex];
    if (note && note.pitchMidi < lowestMidi) {
      lowestMidi = note.pitchMidi;
    }
  }
  if (!Number.isFinite(lowestMidi)) return null;
  const root = parseProgressionChordRoot(label);
  if (root == null) return null;
  return rootMidiBelow(root, lowestMidi);
};

const sequentialCompletedPitchClasses = (
  chord: DefensePhraseChord,
  step: PhraseChordStep,
  correctNoteIndices: ReadonlySet<number>,
): number[] => {
  const midis = stepMidiNotes(chord, step);
  const orderedPcs = orderedPitchClassesFromMidis(midis);
  const completed: number[] = [];
  for (const pc of orderedPcs) {
    const matched = step.noteIndices.some((noteIndex) => {
      const note = chord.notes[noteIndex];
      return Boolean(
        note
        && correctNoteIndices.has(noteIndex)
        && normalizePitchClass(note.pitchClass) === pc,
      );
    });
    if (!matched) {
      break;
    }
    completed.push(pc);
  }
  return completed;
};

export const evaluateDefensePhraseNoteOn = (
  phrases: readonly DefensePhrase[],
  stageRequiredCompletionCount: number,
  state: DefensePhraseJudgeState,
  pitchClass: number,
  sequential = false,
  attackTrigger: DefenseAttackTrigger = 'note',
  autoAdvance = true,
  playStyle: DefensePlayStyle = 'phrase',
  playRootOnChordChange = false,
): DefensePhraseNoteEvaluation => {
  const phrase = phrases[state.phraseIndex] ?? null;
  const chord = getCurrentChord(phrase, state.chordIndex);
  if (!phrase || !chord || chord.notes.length === 0) {
    return {
      attack: false,
      phraseCompleted: false,
      measureCompleted: false,
      pendingSwitch: state.pendingSwitch,
      completionCount: state.completionCount,
      playRootMidi: null,
      nextState: state,
    };
  }

  const { steps } = getPhraseChordSteps(chord.notes);
  const currentStep = steps[state.targetStepIndex];
  if (sequential && currentStep) {
    const acceptance = shouldAcceptChordPitchClassInput(
      stepMidiNotes(chord, currentStep),
      sequentialCompletedPitchClasses(chord, currentStep, state.correctNoteIndices),
      pitchClass,
      true,
    );
    if (!acceptance.accept) {
      return {
        attack: false,
        phraseCompleted: false,
        measureCompleted: false,
        pendingSwitch: state.pendingSwitch,
        completionCount: state.completionCount,
        playRootMidi: null,
        nextState: state,
      };
    }
  }

  const stepState: ChordStepAdvanceState = {
    targetStepIndex: state.targetStepIndex,
    correctNoteIndices: state.correctNoteIndices,
    revealedNoteIndices: state.revealedNoteIndices,
  };

  const evaluation = advanceChordStep(chord.notes, steps, stepState, pitchClass);
  if (evaluation.result === 'miss' || evaluation.result === 'chord-hold') {
    return {
      attack: false,
      phraseCompleted: false,
      measureCompleted: false,
      pendingSwitch: state.pendingSwitch,
      completionCount: state.completionCount,
      playRootMidi: null,
      nextState: state,
    };
  }

  const progressedState: DefensePhraseJudgeState = {
    ...applyStepState(state, evaluation.nextState),
    lastAcceptedPitchClass: pitchClass,
  };
  const isVoicingMode = playStyle === 'chord_voicing';
  const completedStepIndex = evaluation.result === 'measure-complete'
    ? Math.max(0, steps.length - 1)
    : state.targetStepIndex;
  const stepAdvanced = evaluation.result === 'measure-complete'
    || evaluation.nextState.targetStepIndex > state.targetStepIndex;
  const playRootMidi = stepAdvanced
    ? resolvePlayRootMidi(chord, completedStepIndex, playRootOnChordChange)
    : null;
  const attackOnStep = isVoicingMode
    ? stepAdvanced
    : (evaluation.result === 'measure-complete' || attackTrigger === 'note');

  if (evaluation.result === 'measure-complete') {
    const afterChord = advanceChord(progressedState, phrase);
    const wrapped = afterChord.chordIndex === 0;
    if (wrapped) {
      const required = requiredCompletionCountForPhrase(phrase, stageRequiredCompletionCount);
      const nextCount = state.pendingSwitch
        ? state.completionCount
        : state.completionCount + 1;
      const pendingSwitch = autoAdvance && (nextCount >= required || state.pendingSwitch);
      return {
        attack: attackOnStep,
        phraseCompleted: true,
        measureCompleted: true,
        pendingSwitch,
        completionCount: nextCount,
        playRootMidi,
        nextState: {
          ...afterChord,
          completionCount: nextCount,
          pendingSwitch,
        },
      };
    }
    return {
      attack: attackOnStep,
      phraseCompleted: false,
      measureCompleted: true,
      pendingSwitch: state.pendingSwitch,
      completionCount: state.completionCount,
      playRootMidi,
      nextState: {
        ...afterChord,
        completionCount: state.completionCount,
        pendingSwitch: state.pendingSwitch,
      },
    };
  }

  return {
    attack: attackOnStep,
    phraseCompleted: false,
    measureCompleted: false,
    pendingSwitch: state.pendingSwitch,
    completionCount: state.completionCount,
    playRootMidi,
    nextState: {
      ...progressedState,
      completionCount: state.completionCount,
      pendingSwitch: state.pendingSwitch,
    },
  };
};

const isOnPhraseLoopBoundary = (
  phrase: DefensePhrase,
  state: DefensePhraseJudgeState,
): boolean => {
  if (phrase.chords.length === 0) {
    return false;
  }
  const lastChordIndex = phrase.chords.length - 1;
  if (state.chordIndex !== lastChordIndex) {
    return false;
  }
  const lastChord = phrase.chords[lastChordIndex];
  if (!lastChord) {
    return false;
  }
  const { steps } = getPhraseChordSteps(lastChord.notes);
  return state.targetStepIndex >= steps.length - 1;
};

export const getDefenseExpectedPitchCandidates = (
  phrases: readonly DefensePhrase[],
  state: DefensePhraseJudgeState,
  sequential: boolean,
): ExpectedPitchCandidates => {
  if (!sequential) {
    return buildExpectedPitchCandidates([]);
  }
  const hints = getDefensePhraseKeyboardHints(phrases, state, sequential);
  const midis: number[] = [];
  if (hints.nextMidi != null) {
    midis.push(hints.nextMidi);
  }
  const phrase = phrases[state.phraseIndex];
  if (phrase && isOnPhraseLoopBoundary(phrase, state)) {
    const firstChord = phrase.chords[0];
    const firstSteps = firstChord ? getPhraseChordSteps(firstChord.notes).steps : [];
    const firstStep = firstSteps[0];
    if (firstChord && firstStep) {
      const loopStart = computeOrderedChordKeyboardHintsFromMidis(
        stepMidiNotes(firstChord, firstStep),
        [],
      );
      if (loopStart.nextMidi != null && !midis.includes(loopStart.nextMidi)) {
        midis.push(loopStart.nextMidi);
      }
    }
  }
  const repeatPitchClassMask = isDefenseWaitingForSamePitchRepeat(phrases, state, sequential)
    && state.lastAcceptedPitchClass !== null
    ? 1 << state.lastAcceptedPitchClass
    : 0;
  return buildExpectedPitchCandidates(midis, repeatPitchClassMask);
};

export const getDefensePhraseKeyboardHints = (
  phrases: readonly DefensePhrase[],
  state: DefensePhraseJudgeState,
  sequential: boolean,
): OrderedChordKeyboardHints => {
  const empty: OrderedChordKeyboardHints = {
    nextMidi: null,
    pendingMidis: [],
    completedMidis: [],
  };
  const phrase = phrases[state.phraseIndex] ?? null;
  const chord = getCurrentChord(phrase, state.chordIndex);
  if (!phrase || !chord) return empty;

  const { steps } = getPhraseChordSteps(chord.notes);
  const step = steps[state.targetStepIndex];
  if (!step) return empty;

  const midis = stepMidiNotes(chord, step);
  if (midis.length === 0) return empty;

  if (sequential) {
    return computeOrderedChordKeyboardHintsFromMidis(
      midis,
      sequentialCompletedPitchClasses(chord, step, state.correctNoteIndices),
    );
  }

  const pendingMidis: number[] = [];
  const completedMidis: number[] = [];
  for (const noteIndex of step.noteIndices) {
    const note = chord.notes[noteIndex];
    if (!note) continue;
    if (state.correctNoteIndices.has(noteIndex)) {
      completedMidis.push(note.pitchMidi);
    } else {
      pendingMidis.push(note.pitchMidi);
    }
  }
  return { nextMidi: null, pendingMidis, completedMidis };
};

export const nextPhraseIndex = (
  phrases: readonly DefensePhrase[],
  currentIndex: number,
): number => {
  if (phrases.length === 0) return 0;
  return (currentIndex + 1) % phrases.length;
};

File: ios/Jazzify/Defense/Engine/DefensePhraseJudge.swift
import Foundation

enum DefensePhraseJudge {
    struct Evaluation: Equatable {
        let attack: Bool
        let phraseCompleted: Bool
        let measureCompleted: Bool
        let pendingSwitch: Bool
        let completionCount: Int
        let playRootMidi: Int?
        let nextState: DefensePhraseJudgeState
    }

    static func createInitialState(phrases: [DefensePhraseDefinition], phraseIndex: Int = 0) -> DefensePhraseJudgeState {
        DefensePhraseJudgeState(
            phraseIndex: phraseIndex,
            chordIndex: 0,
            targetStepIndex: 0,
            correctNoteIndices: [],
            revealedNoteIndices: [],
            completionCount: 0,
            pendingSwitch: false,
            phrases: phrases
        )
    }

    struct KeyboardHints: Equatable {
        var nextMidis: Set<Int>
        var pendingMidis: Set<Int>
        var completedMidis: Set<Int>

        var allMidis: Set<Int> {
            nextMidis.union(pendingMidis).union(completedMidis)
        }
    }

    static func evaluateNoteOn(
        state: DefensePhraseJudgeState,
        stageRequiredCompletionCount: Int,
        pitchClass: Int,
        sequential: Bool = false,
        attackTrigger: DefenseAttackTrigger = .note,
        autoAdvance: Bool = true,
        playStyle: DefensePlayStyle = .phrase,
        playRootOnChordChange: Bool = false
    ) -> Evaluation {
        guard let phrase = state.phrases[safe: state.phraseIndex],
              let chord = phrase.chords[safe: state.chordIndex],
              !chord.notes.isEmpty
        else {
            return Evaluation(
                attack: false,
                phraseCompleted: false,
                measureCompleted: false,
                pendingSwitch: state.pendingSwitch,
                completionCount: state.completionCount,
                playRootMidi: nil,
                nextState: state
            )
        }

        let steps = SurvivalPhraseChordSteps.getSteps(notes: chord.notes)
        if sequential,
           let currentStep = steps[safe: state.targetStepIndex] {
            let midis = stepMidis(chord: chord, step: currentStep)
            let completed = sequentialCompletedPitchClasses(
                chord: chord,
                step: currentStep,
                correctNoteIndices: state.correctNoteIndices
            )
            let nextPc = SurvivalChordResolver.nextExpectedPitchClass(
                fromMidis: midis,
                inputPitchClasses: completed
            )
            let pc = ((pitchClass % 12) + 12) % 12
            if nextPc != pc {
                return Evaluation(
                    attack: false,
                    phraseCompleted: false,
                    measureCompleted: false,
                    pendingSwitch: state.pendingSwitch,
                    completionCount: state.completionCount,
                    playRootMidi: nil,
                    nextState: state
                )
            }
        }

        let stepState = SurvivalPhraseChordSteps.AdvanceState(
            targetStepIndex: state.targetStepIndex,
            correctNoteIndices: state.correctNoteIndices,
            revealedNoteIndices: state.revealedNoteIndices
        )
        let evaluation = SurvivalPhraseChordSteps.advance(
            notes: chord.notes,
            steps: steps,
            state: stepState,
            pitchClass: pitchClass
        )

        switch evaluation.result {
        case .miss, .chordHold:
            return Evaluation(
                attack: false,
                phraseCompleted: false,
                measureCompleted: false,
                pendingSwitch: state.pendingSwitch,
                completionCount: state.completionCount,
                playRootMidi: nil,
                nextState: state
            )
        case .progress, .measureComplete:
            break
        }

        var next = state
        next.targetStepIndex = evaluation.nextState.targetStepIndex
        next.correctNoteIndices = evaluation.nextState.correctNoteIndices
        next.revealedNoteIndices = evaluation.nextState.revealedNoteIndices

        let isVoicingMode = playStyle == .chordVoicing
        let completedStepIndex = evaluation.result == .measureComplete
            ? max(0, steps.count - 1)
            : state.targetStepIndex
        let stepAdvanced = evaluation.result == .measureComplete
            || evaluation.nextState.targetStepIndex > state.targetStepIndex
        let playRootMidi = stepAdvanced
            ? resolvePlayRootMidi(
                chord: chord,
                stepIndex: completedStepIndex,
                playRootOnChordChange: playRootOnChordChange
            )
            : nil
        let attackOnStep = isVoicingMode
            ? stepAdvanced
            : (evaluation.result == .measureComplete || attackTrigger == .note)

        if evaluation.result == .measureComplete {
            next = advanceChord(next, phrase: phrase)
            if next.chordIndex == 0 {
                let required = phrase.requiredCompletionCount ?? stageRequiredCompletionCount
                let nextCount = state.pendingSwitch ? state.completionCount : state.completionCount + 1
                let pending = autoAdvance && (nextCount >= required || state.pendingSwitch)
                next.completionCount = nextCount
                next.pendingSwitch = pending
                return Evaluation(
                    attack: attackOnStep,
                    phraseCompleted: true,
                    measureCompleted: true,
                    pendingSwitch: pending,
                    completionCount: nextCount,
                    playRootMidi: playRootMidi,
                    nextState: next
                )
            }
            return Evaluation(
                attack: attackOnStep,
                phraseCompleted: false,
                measureCompleted: true,
                pendingSwitch: state.pendingSwitch,
                completionCount: state.completionCount,
                playRootMidi: playRootMidi,
                nextState: next
            )
        }

        return Evaluation(
            attack: attackOnStep,
            phraseCompleted: false,
            measureCompleted: false,
            pendingSwitch: state.pendingSwitch,
            completionCount: state.completionCount,
            playRootMidi: playRootMidi,
            nextState: next
        )
    }

    static func resetToPhraseIndex(_ index: Int, phrases: [DefensePhraseDefinition]) -> DefensePhraseJudgeState {
        createInitialState(phrases: phrases, phraseIndex: index)
    }

    static func targetMidis(state: DefensePhraseJudgeState) -> [Int] {
        Array(keyboardHints(state: state, sequential: false).pendingMidis)
    }

    static func keyboardHints(state: DefensePhraseJudgeState, sequential: Bool) -> KeyboardHints {
        let empty = KeyboardHints(nextMidis: [], pendingMidis: [], completedMidis: [])
        guard let phrase = state.phrases[safe: state.phraseIndex],
              let chord = phrase.chords[safe: state.chordIndex]
        else { return empty }
        let steps = SurvivalPhraseChordSteps.getSteps(notes: chord.notes)
        guard let step = steps[safe: state.targetStepIndex] else { return empty }
        let midis = stepMidis(chord: chord, step: step)
        if sequential {
            let completedPcs = sequentialCompletedPitchClasses(
                chord: chord,
                step: step,
                correctNoteIndices: state.correctNoteIndices
            )
            let nextPc = SurvivalChordResolver.nextExpectedPitchClass(
                fromMidis: midis,
                inputPitchClasses: completedPcs
            )
            var nextMidis = Set<Int>()
            var pending = Set<Int>()
            var completed = Set<Int>()
            let ordered = SurvivalChordResolver.orderedPitchClasses(fromMidis: midis)
            let completedSet = Set(completedPcs)
            for midi in midis {
                let pc = ((midi % 12) + 12) % 12
                if completedSet.contains(pc) {
                    completed.insert(midi)
                } else if pc == nextPc {
                    nextMidis.insert(midi)
                } else if ordered.contains(pc) {
                    pending.insert(midi)
                }
            }
            return KeyboardHints(nextMidis: nextMidis, pendingMidis: pending, completedMidis: completed)
        }

        var pending = Set<Int>()
        var completed = Set<Int>()
        for index in step.noteIndices {
            guard let note = chord.notes[safe: index] else { continue }
            if state.correctNoteIndices.contains(index) {
                completed.insert(note.pitchMidi)
            } else {
                pending.insert(note.pitchMidi)
            }
        }
        return KeyboardHints(nextMidis: [], pendingMidis: pending, completedMidis: completed)
    }

    static func nextPhraseIndex(phrases: [DefensePhraseDefinition], current: Int) -> Int {
        guard !phrases.isEmpty else { return 0 }
        return (current + 1) % phrases.count
    }

    private static func resolvePlayRootMidi(
        chord: SurvivalPhraseChord,
        stepIndex: Int,
        playRootOnChordChange: Bool
    ) -> Int? {
        guard playRootOnChordChange else { return nil }
        let label = DefenseStaffChordLabel.label(for: chord, stepIndex: stepIndex)
        guard !label.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return nil }
        let steps = SurvivalPhraseChordSteps.getSteps(notes: chord.notes)
        guard let step = steps[safe: stepIndex] else { return nil }
        var lowestMidi = Int.max
        for noteIndex in step.noteIndices {
            guard let note = chord.notes[safe: noteIndex] else { continue }
            lowestMidi = min(lowestMidi, note.pitchMidi)
        }
        guard lowestMidi != Int.max else { return nil }
        guard let root = TrainingProgression.parseProgressionChordRoot(label) else { return nil }
        return TrainingMusicTheory.rootMidiBelow(root: root, lowestMidi: lowestMidi)
    }

    private static func advanceChord(
        _ state: DefensePhraseJudgeState,
        phrase: DefensePhraseDefinition
    ) -> DefensePhraseJudgeState {
        var next = state
        guard !phrase.chords.isEmpty else { return next }
        next.chordIndex = (state.chordIndex + 1) % phrase.chords.count
        next.targetStepIndex = 0
        next.correctNoteIndices = []
        next.revealedNoteIndices = []
        return next
    }

    private static func stepMidis(chord: SurvivalPhraseChord, step: PhraseChordStep) -> [Int] {
        step.noteIndices.compactMap { chord.notes[safe: $0]?.pitchMidi }
    }

    private static func sequentialCompletedPitchClasses(
        chord: SurvivalPhraseChord,
        step: PhraseChordStep,
        correctNoteIndices: Set<Int>
    ) -> [Int] {
        let midis = stepMidis(chord: chord, step: step)
        let ordered = SurvivalChordResolver.orderedPitchClasses(fromMidis: midis)
        var completed: [Int] = []
        for pc in ordered {
            let matched = step.noteIndices.contains { index in
                guard let note = chord.notes[safe: index] else { return false }
                return correctNoteIndices.contains(index)
                    && ((note.pitchClass % 12) + 12) % 12 == pc
            }
            if !matched { break }
            completed.append(pc)
        }
        return completed
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        guard indices.contains(index) else { return nil }
        return self[index]
    }
}
